twill-yarn/src/main/java/org/apache/twill/internal/appmaster/RunningContainers.java

```diff
@@ -477,10 +493,30 @@ void handleCompleted(YarnContainerStatus status, Multiset<String> restartRunnabl
     }
   }
 
-  private boolean shouldRetry(int exitCode) {
-    return exitCode != ContainerExitCodes.SUCCESS
-      && exitCode != ContainerExitCodes.DISKS_FAILED
-      && exitCode != ContainerExitCodes.INIT_FAILED;
+  private boolean shouldRetry(String runnableName, int instanceId, int exitCode) {
+    boolean possiblyRetry = 
+        exitCode != ContainerExitCodes.SUCCESS && 
+        exitCode != ContainerExitCodes.DISKS_FAILED && 
+        exitCode != ContainerExitCodes.INIT_FAILED;
+    
+    if (possiblyRetry) {
+      int max = getMaxRetries(runnableName);
+      if (max == Integer.MAX_VALUE) {
+        return true; // retry without special log msg
+      }
+
+      if (getRetryCount(runnableName, instanceId) == max) {
```

Comment:
Since we call `getRetryCount` for this check, might as well cache it as local var to make sure we got  the right one for this request:

```
int retryCount = getRetryCount(runnableName, instanceId);
if (getRetryCount(runnableName, instanceId) == max) {
  ...
} else {
  LOG.info("Attempting {} of {} retries for instance {} of runnable {}.", retryCount + 1,
    max, instanceId, runnableName);
  return true;
}
```